### DISCUSSION TITLE
中英切换是否可以默认使用 shortCommand 完成，而非特制的英文键盘

### DISCUSSION THREAD

**Discussion by tansongchen** [imfuxiao/Hamster#434] [Ideas]
目前，有大量的 issue 都是关于中英切换后键盘布局不一致的：如 #391 #397 #401 等等。在制作声笔键盘的过程中，我使用的是 `shortCommand(#中英切换)`，就没有这类问题。

很多 Rime 用户都在电脑上有使用 Rime 的经验，对于使用 Rime 本身切换中英文的效果是比较熟悉的；同文输入法更是把 ascii_mode 的切换做到了候选栏上，与电脑的行为是一致的。是否有可能将默认的中英切换改成 `shortCommand(#中英切换)`，如果用户需要进一步定制英文的输入体验，再去自己定制英文键盘？

如果默认是这样的行为，那么我觉得可能用户产生的疑惑会少一些。

**Comment by imfuxiao** [comment]
RIME 的 ascii 模式 在不同的输入方案下，表现不一致。比如有些固定码长的输入方案，只能输入4个英文符号。这对于 RIME 不熟悉的用户，这种行为会有困扰。所以不建议将 ascii 模式作为默认中英切换模式。

**Reply by groverlynn** [reply]
Rime的ASCII composer兜底的composer，輸入方案的speller模塊無法影響（截胡）優先級比自己高的模塊的。所以Rime任何輸入方案切換ASCII以後行為是完全一樣的。除非ASCII composer的加載順序被調後了。固定碼長只是針對輸入方案自身，輸入標點符號、反查、pattern等等都不受限制。

**Reply by imfuxiao** [reply]
抱歉，我的问题。在 mac 上把 `ascii_composer` 禁用了。原因是 鼠须管-中文，鼠须管-英文，系统英文，三个切换想改为两个。

https://github.com/rime/squirrel/issues/221#issuecomment-401652387

禁用后就会是您说的没有 `ascii_composer` 兜底了。

**Comment by tansongchen** [comment]
在 Rime 的引擎中，`ascii_mode` 这个开关是由 `ascii_composer` 响应的，逻辑在[这里](https://github.com/rime/librime/blob/767ebad054588ec2b99433cef6922283042de2fe/src/rime/gear/ascii_composer.cc#L113)；即，如果 `ascii_mode = 1`，任何常规按键都会返回 `kRejected`。固定码长属于 `speller` 引擎的配置，而配置文件中 `ascii_composer` 约定俗成是优先级最高的，所以应该不会有问题，如果优先级不对的话应该是方案作者的问题。

换个角度想，在鼠须管等电脑平台上，默认安装之后就是启用了 Shift 切换 `ascii_mode`，大多数人其实也是用 Shift 去切换，不会真正用 Control+Space 切换到英文键盘去输入。同文上也把 `ascii_mode` 直接放到了候选栏上让用户去切换，也都没有人反馈过有打不出英文符号的问题，说明 `ascii_composer` 的优先级是被大多数 Rime 方案作者和用户遵守的。所以如果仓也像这些平台一样认为 `ascii_mode` 即是纯英文输入，应该不至于产生什么问题。

不知道这么解释是不是回答了您的疑问，如果我说的不符合实际情况的话，您可以发我一个使用 `ascii_mode` 打英文会出现问题的示例我研究研究。

**Reply by imfuxiao** [reply]
您好，我用的小鹤音形方案 http://flypy.ysepan.com/ 切换 ascii 模式后就会存在这个现象。

**Reply by tansongchen** [reply]
在我的鼠须管上没能成功复现这个问题……目前先搁置吧，以后再研究一下。

**Reply by imfuxiao** [reply]
我也是鼠须管，它的配置模式使用 ctrl + shift + 2 切换 ascii 模式。

**Comment by g6ai** [comment]
请问九宫格键盘能关闭英文键盘吗，我想在九宫格和系统英文键盘之间切换

**Reply by imfuxiao** [reply]
抱歉，当前版本不支持中文九宫格的布局自定义。

**Comment by 23fo** [comment]
使用`shortCommand(#中英切换)`会带来另一个问题，我使用仓颉键盘，切换后键盘不是英文字母也不太舒适，使用注音键盘的应该会有类似问题
所以能否支持自定义英文键盘呢

**Reply by imfuxiao** [reply]
有内置英文键盘，可以直接切换到内置英文键盘，具体看下 wiki 中的自定义键盘文档。

> 注意：商店版本的「自定义键盘」文档与TF目前测试中的「自定义键盘（alpha）」是不一样的，预计本周会发布。

**Reply by 23fo** [reply]
内置的英文键盘就有 @tansongchen 说的问题，键盘切换键必须与内置的英文键盘保持一致，且有键尺寸不统一的问题
![img_v3_026b_aa8ac87a-7b52-435d-ba5c-d77cc8f23eag](https://github.com/imfuxiao/Hamster/assets/3532135/1724d4bf-3a82-43b7-a016-b978ffd46caa)
![img_v3_026b_c14424c8-d3ba-4fd5-876a-cbd4c487c05g](https://github.com/imfuxiao/Hamster/assets/3532135/57dc055a-e76d-4d2d-8573-13f3dd70e350)

**Reply by imfuxiao** [reply]
您好，尺寸不一致的问题我不是很明白，还请您详细说明。

另外键盘切换键与内置英文键盘保持一致的问题，内置的仓颉键盘本身就是自定义键盘，您可以直接修改配置文件。

**Reply by 23fo** [reply]
1. 尺寸可以看图中，自定义键盘高度比内置因为键盘低；
2. 键盘切换键我知道可以修改，但只能修改自定义键盘去适应内置键盘，这个逻辑是不对的；
基于以上两点都是用户想要好的体验都要用自定义键盘去适应内置键盘，就失去了自定义键盘的部分意义，所以才提议自定义英文键盘

**Reply by imfuxiao** [reply]
1. 高度的问题我真没看出来，按照内置的自定义仓颉键盘参数来说，它用的高度与内置键盘的高度是一致的，因为高度参数为空，就表示使用内置的高度。
2. 这个还是看个个人偏好吧。不喜内置英文键盘的化，可以自定义的。那种都行。

**Reply by 23fo** [reply]
现在支持自定义英文键盘吗？我是看了WIKI好像不能才来提议的

**Reply by imfuxiao** [reply]
肯定是可以的，为啥觉的不行能，麻烦具体说明问题，我才可以帮助处理。

**Reply by 23fo** [reply]
自定义键盘没有指定输入法的参数，或者说切换键盘的时候没有同时切换输入的参数，切换到英文键盘输入法没有变化

**Reply by imfuxiao** [reply]
1. 目前英文键盘的处理逻辑是按键不经过 Rime 引擎，按键配置什么，就上屏什么，内置键盘就是这样的。缺点：无法使用英文输入方案。
2. 如您所说，键盘布局目前与输入方案是没有关联关系的，需要自己切换，这个功能还在考虑中

**Reply by 23fo** [reply]
多套输入方案需求量应该比较小，或者可以考虑把英文键盘布局开放出来

**Comment by ZmdnaHF0ZmFqYXZ0O** [comment]
请问中英切换后的英文键盘，需要上滑按键输入字符可以修改哪个文件实现？
